lrpar/src/lib/ctbuilder.rs

```diff
@@ -754,57 +753,69 @@ where
-        // Emit user program section, and actions at the top so they may specify inner attributes.
-        if let Some(YaccKind::Original(YaccOriginalActionKind::UserAction) | YaccKind::Grmtools) =
-            self.yacckind
+        let user_actions = if let Some(
+            YaccKind::Original(YaccOriginalActionKind::UserAction) | YaccKind::Grmtools,
+        ) = self.yacckind
         {
-            outs.push_str(&self.gen_user_actions(grm)?);
-        }
-        outs.push_str("    mod _parser_ {\n");
-        outs.push_str(
-            "        #![allow(clippy::type_complexity)]
-        #![allow(clippy::unnecessary_wraps)]
-        #![deny(unsafe_code)]
-        #[allow(unused_imports)]
-        use super::*;
-",
-        );
+            Some(str::parse::<TokenStream>(&self.gen_user_actions(grm)?).unwrap())
+        } else {
+            None
+        };
 
-        outs.push_str(&self.gen_parse_function(grm, stable)?);
-        outs.push_str(&self.gen_rule_consts(grm));
-        outs.push_str(&self.gen_token_epp(grm));
-        match self.yacckind.unwrap() {
+        let rule_consts = str::parse::<TokenStream>(&self.gen_rule_consts(grm)).unwrap();
+        let token_epp = str::parse::<TokenStream>(&self.gen_token_epp(grm)).unwrap();
+        let parse_function = self.gen_parse_function(grm, stable)?;
+        let action_wrappers = match self.yacckind.unwrap() {
             YaccKind::Original(YaccOriginalActionKind::UserAction) | YaccKind::Grmtools => {
-                outs.push_str(&self.gen_wrappers(grm));
+                Some(str::parse::<TokenStream>(&self.gen_wrappers(grm)).unwrap())
             }
             YaccKind::Original(YaccOriginalActionKind::NoAction)
-            | YaccKind::Original(YaccOriginalActionKind::GenericParseTree) => (),
+            | YaccKind::Original(YaccOriginalActionKind::GenericParseTree) => None,
             _ => unreachable!(),
-        }
-        outs.push_str("    } // End of `mod _parser_`\n\n");
-        outs.push_str("    #[allow(unused_imports)]\n");
-        outs.push_str("    pub use _parser_::*;\n");
-        outs.push_str("    #[allow(unused_imports)]\n");
-        outs.push_str("    use ::lrpar::Lexeme;\n");
-
-        // Output the cache so that we can check whether the IDs map is stable.
-        outs.push_str(cache);
-        outs.push_str("} // End of `mod {mod_name}` \n\n");
+        };
 
+        let mod_name = format_ident!("{}", mod_name);
+
+        let out_tokens = quote! {
+            #visibility mod #mod_name {
+                #cache
+                // gen_user_actions
+                #user_actions
+
+                mod _parser_ {
+                    #![allow(clippy::type_complexity)]
+                    #![allow(clippy::unnecessary_wraps)]
+                    #![deny(unsafe_code)]
+                    #[allow(unused_imports)]
+                    use super::*;
+                } // End of `mod _parser_`
+
+                #parse_function
```

Comment:
Doesn't look right, I think `#parse_function` is supposed to be moved up into the `_parser_` module.
This also explains why I had to add `clippy::type_complexity` to the `parse_function`.